### PR TITLE
REF: Refactor Date/TimeLikeOps

### PR DESCRIPTION
**Pull request by TomAugspurger**
No real functional changes, just an inheritance reorganization to make the diff at #24024 smaller.

Changes:

- Removes DatelikeOps from PeriodIndex (already had strftime via delegation)
- Moves Date/TimelikeOps from DatetimeIndex to DatetimeArray
- Moves TimelikeOps from TimedeltaIndex to TimedeltaArray

### CODE REVIEW

**Comment by jbrockmendel** [comment]
+1 on the scope.  Needs tests for the methods added to DTA/TDA/PA

**Comment by TomAugspurger** [comment]
In response to 
https://github.com/pandas-dev/pandas/pull/24038#pullrequestreview-180544284, 
https://github.com/pandas-dev/pandas/pull/24038#pullrequestreview-180544274, and 
https://github.com/pandas-dev/pandas/pull/24038#pullrequestreview-180544255, nothing has changed here. These classes end up mixed into exactly the same set as before. This is merely to keep the diff at #24024 smaller.

So in response to

> Needs tests for the methods added to DTA/TDA/PA

That can't be done yet (at least I don't see an easy way), until the indexes have been split from the arrays.

File: pandas/core/arrays/datetimelike.py [hunk 1]
```diff
@@ -1023,3 +1208,39 @@ def validate_dtype_freq(dtype, freq):
...
+    other : 1d array
+    to_utc : bool, default False
+        If True, convert the values to UTC before extracting the i8 values
+        If False, extract the i8 values directly.
+
+    Returns
+    -------
+    i8 1d array
+    """
+    from pandas import Index
```
[the remark this answers is not in the document]
Comment (TomAugspurger) [reply pandas/core/arrays/datetimelike.py]: That's what the old version did. I haven't looked at which is more appropriate here.

**Comment by jreback** [comment]
ok with going ahead with this to avoid blocking and making diffs easier. @jbrockmendel

**Comment by jbrockmendel** [comment]
OK.  I'll see if I can implement tests in a follow-up.

**Comment by jreback** [comment]
thanks @TomAugspurger

**Pull request merged**